examples/vtable/realm_bootstrap.cc

```diff
@@ -0,0 +1,171 @@
+
+// Realm NetworkVtable required keys
+static constexpr const char *REALM_KEY_RANK = "realm_rank";
+static constexpr const char *REALM_KEY_RANKS = "realm_ranks";
+static constexpr const char *REALM_KEY_GROUP = "realm_group";
+
+static std::map<std::string, std::vector<uint8_t>> local_kv_store;
+static std::map<std::string, std::vector<uint8_t>> global_kv_store;
+static bool pending_sync = false;
+
+static bool app_put(const void *key, size_t key_size, const void *value, size_t value_size,
+                    const void *vtable_data, size_t vtable_data_size)
+{
+  if(!key || !value || key_size == 0) return false;
+  std::string k(static_cast<const char*>(key), key_size);
+  std::vector<uint8_t> v(static_cast<const uint8_t*>(value), 
+                         static_cast<const uint8_t*>(value) + value_size);
+  local_kv_store[k] = v;
+  pending_sync = true;
+  // std::cout << "app_put: key='" << k << "' value_size=" << value_size << std::endl;
+  return true;
+}
+
+static bool app_get(const void *key, size_t key_size, void *value, size_t *value_size,
+                    const void *vtable_data, size_t vtable_data_size)
+{
+  if(!key || !value || !value_size || key_size == 0) return false;
+  std::string k(static_cast<const char*>(key), key_size);
+  // std::cout << "app_get: key='" << k << "'" << std::endl;
+  
+  if(k == REALM_KEY_RANK) {
+    int rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    if(*value_size < sizeof(uint32_t)) { *value_size = 0; return false; }
+    uint32_t v = static_cast<uint32_t>(rank);
+    memcpy(value, &v, sizeof(v));
+    *value_size = sizeof(v);
+    return true;
+  }
+  if(k == REALM_KEY_RANKS) {
+    int size;
+    MPI_Comm_size(MPI_COMM_WORLD, &size);
+    if(*value_size < sizeof(uint32_t)) { *value_size = 0; return false; }
+    uint32_t v = static_cast<uint32_t>(size);
+    memcpy(value, &v, sizeof(v));
+    *value_size = sizeof(v);
+    return true;
+  }
+  if(k == REALM_KEY_GROUP) {
+    if(*value_size < sizeof(uint32_t)) { *value_size = 0; return false; }
+    uint32_t v = 0;
+    memcpy(value, &v, sizeof(v));
+    *value_size = sizeof(v);
+    return true;
+  }
+  
+  auto it = global_kv_store.find(k);
+  if(it != global_kv_store.end()) {
+    if(it->second.size() > *value_size) {
+      *value_size = it->second.size();
+      return false;
+    }
+    memcpy(value, it->second.data(), it->second.size());
+    *value_size = it->second.size();
+    return true;
+  }
+  
+  it = local_kv_store.find(k);
+  if(it != local_kv_store.end()) {
+    if(it->second.size() > *value_size) {
+      *value_size = it->second.size();
+      return false;
+    }
+    memcpy(value, it->second.data(), it->second.size());
+    *value_size = it->second.size();
+    return true;
+  }
+  
+  *value_size = 0;
+  return true;
+}
+
+static bool app_bar(const void *vtable_data, size_t vtable_data_size)
+{
+  // std::cout << "app_bar called" << std::endl;
+  
+  if(MPI_Barrier(MPI_COMM_WORLD) != MPI_SUCCESS)
+    return false;
```

Comment:
I don't think you actually need this, the blocking all-gather call farther down also functions as a barrier across the processes because it does not return until it receives something from every other process which means that everyone has arrived at it.